src/ripple/rpc/impl/NFTokenID.cpp

```diff
@@ -0,0 +1,192 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/rpc/NFTokenID.h>
+
+#include <ripple/app/ledger/LedgerMaster.h>
+#include <ripple/app/ledger/OpenLedger.h>
+#include <ripple/app/misc/Transaction.h>
+#include <ripple/ledger/View.h>
+#include <ripple/net/RPCErr.h>
+#include <ripple/protocol/AccountID.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/rpc/Context.h>
+#include <ripple/rpc/impl/RPCHelpers.h>
+#include <boost/algorithm/string/case_conv.hpp>
+
+namespace ripple {
+namespace RPC {
+
+bool
+canHaveNFTokenID(
+    std::shared_ptr<STTx const> const& serializedTx,
+    TxMeta const& transactionMeta)
+{
+    if (!serializedTx)
+        return false;
+
+    TxType const tt{serializedTx->getTxnType()};
```

Comment:
just to confirm, using braces initialization on integers allows _less_ type conversions, and implicit conversions won't work with them? But using braces initialization for objects actually does more conversions?

Reply:
Sadly, yes.  This is one where the C++ Standards Committee got their signals crossed.  Different folks had different priorities for braced initialization.  It was supposed to be a panacea, but just turned into a mess.

First let's talk about integers.  Actually, we're talking about built in numeric types since floats and doubles are also affected.  Braced initialization, as introduced in C++11, explicitly disallows _narrowing conversions_.  Due to its C heritage, C++ sometimes plays fast and loose when assigning one kind of integer to another.  You can, for example assign an integer value to a char.  The compiler will not raise an objection.  There are rules for what the narrowing conversion does; in some cases it lops off the upper bits.  But you cannot brace initialize a char with an integer.  The compiler will stop you because that is a narrowing conversion.

Disallowing narrowing conversions, that's a really good thing.

However, once we move from built in numeric types to more interesting types (with constructors) the committee got too excited and made it _too_ easy to initialize with braces.  I don't remember the whole story, but at least one of the downsides is that braced initialization considers both non-`explicit` and `explicit` constructors.  The _goal_ of introducing `explicit` constructors was to make it hard to _accidentally_ construct a type from the given argument(s).  Using braced initialization throws away the work of developers marking constructors as `explicit`; it makes it _easy_ to use an `explicit` constructor by accident.

So, yeah, unfortunately it's a complicated story.  I've stopped using braced initialization unless I really want it.  I more commonly come across `explicit` constructors in this line of work.  Narrowing integer conversions happen, but it's not that common.

If you really want to use braced initialization then using `= {...}` (with the equals sign) is safer.  This is called copy-list-initialization, and it is only allowed to call non-`explicit` constructors.  But at that point it may (usually) make sense to skip the braces alltogether.

Now you at least have enough information to make your own choice.  Let me know if that needs further clarification.

Reply:
Thanks!!! Good to learn the whole story, didn't know about this before. I would go with `=` then :-)